File: day_11.rs
use std::io::{stdin, Read};
use std::str::FromStr;

fn main() {
    let mut input = String::new();
    stdin().lock().read_to_string(&mut input).unwrap();
    let seat_map = input.parse::<SeatMap>().unwrap();
    let first_answer = seat_map
        .transform(|seat_map, x, y, seat_state| {
            match seat_state {
                SeatState::Empty => {
                    if seat_map.count_adjacent_seats_of_state(x, y, SeatState::Taken) == 0 {
                        return SeatState::Taken
                    }
                },
                SeatState::Taken => {
                    if seat_map.count_adjacent_seats_of_state(x, y, SeatState::Taken) >= 4 {
                        return SeatState::Empty
                    }
                },
                _ => {}
            }
            seat_state
        })
        .count_seats_of_state(SeatState::Taken);
    println!("first answer: {}", first_answer);
    let second_answer = seat_map
        .transform(|seat_map, x, y, seat_state| {
            match seat_state {
                SeatState::Empty => {
                    if seat_map.count_visible_seats_of_state(x, y, SeatState::Taken) == 0 {
                        return SeatState::Taken
                    }
                },
                SeatState::Taken => {
                    if seat_map.count_visible_seats_of_state(x, y, SeatState::Taken) >= 5 {
                        return SeatState::Empty
                    }
                },
                _ => {}
            }
            seat_state
        })
        .count_seats_of_state(SeatState::Taken);
    println!("second answer: {}", second_answer);
}

#[derive(Debug, Clone)]
struct SeatMap {
    row_width: usize,
    states: Vec<SeatState>,
}

impl SeatMap {
    #[inline(always)]
    fn get(self: &Self, x: usize, y: usize) -> SeatState {
        self.states[y * self.row_width + x]
    }

    #[inline(always)]
    fn get_with_delta(self: &Self, x: usize, y: usize, xd: isize, yd: isize) -> Option<SeatState> {
        let ix = x as isize + xd;
        let iy = y as isize + yd;
        if ix < 0 || iy < 0 {
            return None
        }
        let ux = ix as usize;
        let uy = iy as usize;
        if ux >= self.row_width || uy >= self.height() {
            return None
        }
        Some(self.get(ux, uy))
    }

    #[inline(always)]
    fn set(self: &mut Self, x: usize, y: usize, state: SeatState) {
        self.states[y * self.row_width + x] = state;
    }

    #[inline(always)]
    fn height(self: &Self) -> usize {
        self.states.len() / self.row_width
    }

    fn count_seats_of_state(self: &Self, state: SeatState) -> usize {
        self.states.iter().filter(|s| **s == state).count()
    }

    fn count_visible_seats_of_state(self: &Self, x: usize, y: usize, state: SeatState) -> usize {
        let mut count = 0;
        for yd in -1..=1 {
            for xd in -1..=1 {
                if xd != 0 || yd != 0 {
                    let mut i = 1;
                    'find_non_floor: loop {
                        match self.get_with_delta(x, y, xd * i, yd * i) {
                            Some(s) => {
                                if s == state {
                                    count += 1;
                                } else if s == SeatState::Floor {
                                    i += 1;
                                    continue 'find_non_floor;
                                }
                            },
                            _ => {}
                        }
                        break;
                    }
                }
            }
        }
        count
    }

    fn count_adjacent_seats_of_state(self: &Self, x: usize, y: usize, state: SeatState) -> usize {
        let mut count = 0;
        for yd in -1..=1 {
            for xd in -1..=1 {
                if xd != 0 || yd != 0 {
                    match self.get_with_delta(x, y, xd, yd) {
                        Some(s) => {
                            if s == state {
                                count += 1;
                            }
                        },
                        _ => {}
                    }
                }
            }
        }
        count
    }

    #[allow(dead_code)]
    fn print(self: &Self) {
        print!("{esc}[2J{esc}[1;1H", esc = 27 as char);
        println!("{}\n", self.states.chunks(self.row_width).map(|s| {
            s.iter().map(|s| match s {
                SeatState::Floor => ".",
                SeatState::Empty => "L",
                SeatState::Taken => "#",
            }.to_string()).collect::<Vec<String>>().join("")
        }).collect::<Vec<String>>().join("\n"));
    }

    fn transform(self: &Self, rearrange: fn(&Self, usize, usize, SeatState) -> SeatState) -> Self {
        let mut scratch = [self.clone(), self.clone()];
        let mut dst = 0;
        loop {
            let src = dst;
            dst ^= 1;
            let mut changes = 0;
            for y in 0..self.height() {
                for x in 0..self.row_width {
                    let old = scratch[src].get(x, y);
                    let new = rearrange(&scratch[src], x, y, old);
                    scratch[dst].set(x, y, new);
                    if old != new {
                        changes += 1;
                    }
                }
            }
            if changes == 0 {
                return Self{ row_width: self.row_width, states: scratch[dst].states.to_owned() }
            }
        }
    }
}

#[derive(Debug)]
enum ParseSeatMapError {
    InvalidSeatState(ParseSeatStateError),
}

impl FromStr for SeatMap {
    type Err = ParseSeatMapError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let mut row_width = 0;
        let states = match s
            .lines()
            .flat_map(|line| {
                row_width = line.len();
                line
                    .chars()
                    .map(|c| c.to_string().parse::<SeatState>())
            })
            .collect
            ::<Result<Vec<SeatState>, ParseSeatStateError>>() {
                Ok(s) => s,
                Err(e) => return Err(Self::Err::InvalidSeatState(e))
            };
        Ok(Self{ row_width: row_width, states: states })
    }
}

#[derive(Debug, Clone, Copy, PartialEq)]
enum SeatState {
    Floor,
    Empty,
    Taken,
}

#[derive(Debug)]
enum ParseSeatStateError {
    InvalidState(String),
}

impl FromStr for SeatState {
    type Err = ParseSeatStateError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "." => Ok(Self::Floor),
            "L" => Ok(Self::Empty),
            "#" => Ok(Self::Taken),
            _ => Err(Self::Err::InvalidState(s.to_string()))
        }
    }
}

File: day_11.go
package main

import (
	"bytes"
	"fmt"
	"io"
	"io/ioutil"
	"os"
)

func main() {
	m, err := parseSeatMap(os.Stdin)
	if err != nil {
		panic(err)
	}
	fmt.Println(m.RowWidth, len(m.Seats)/m.RowWidth)
	scratch := [...]seatMap{m.Clone(), m.Clone()}
	firstAnswer := m.RearrangeUntilStable(scratch, func(m seatMap, x, y int, seatType seatType) seatType {
		switch seatType {
		case seatTypeEmpty:
			if m.CountAdjacentSeatsOfType(x, y, seatTypeTaken) == 0 {
				return seatTypeTaken
			}
		case seatTypeTaken:
			if m.CountAdjacentSeatsOfType(x, y, seatTypeTaken) >= 4 {
				return seatTypeEmpty
			}
		}
		return seatType
	}).CountSeatsOfType(seatTypeTaken)
	fmt.Println("first answer:", firstAnswer)
	secondAnswer := m.RearrangeUntilStable(scratch, func(m seatMap, x, y int, seatType seatType) seatType {
		switch seatType {
		case seatTypeEmpty:
			if m.CountVisibleSeatsOfType(x, y, seatTypeTaken) == 0 {
				return seatTypeTaken
			}
		case seatTypeTaken:
			if m.CountVisibleSeatsOfType(x, y, seatTypeTaken) >= 5 {
				return seatTypeEmpty
			}
		}
		return seatType
	}).CountSeatsOfType(seatTypeTaken)
	fmt.Println("second answer:", secondAnswer)
}

type seatMap struct {
	RowWidth int
	Seats    []seatType
}

func parseSeatMap(r io.Reader) (seatMap, error) {
	data, err := ioutil.ReadAll(r)
	if err != nil {
		return seatMap{}, err
	}
	seats := make([]seatType, 0, len(data))
	var rowWidth int
	for _, line := range bytes.Split(data, []byte{'\n'}) {
		rowWidth = len(line)
		for _, c := range line {
			switch c {
			case '.':
				seats = append(seats, seatTypeFloor)
			case 'L':
				seats = append(seats, seatTypeEmpty)
			case '#':
				seats = append(seats, seatTypeTaken)
			}
		}
	}
	return seatMap{rowWidth, seats}, nil
}

func (m seatMap) CountVisibleSeatsOfType(x, y int, seatType seatType) int {
	count := 0
	count += btoi(m.HasSeatVisibleInDirection(x, y, -1, -1, seatType))
	count += btoi(m.HasSeatVisibleInDirection(x, y, +0, -1, seatType))
	count += btoi(m.HasSeatVisibleInDirection(x, y, +1, -1, seatType))
	count += btoi(m.HasSeatVisibleInDirection(x, y, -1, +0, seatType))
	count += btoi(m.HasSeatVisibleInDirection(x, y, +1, +0, seatType))
	count += btoi(m.HasSeatVisibleInDirection(x, y, -1, +1, seatType))
	count += btoi(m.HasSeatVisibleInDirection(x, y, +0, +1, seatType))
	count += btoi(m.HasSeatVisibleInDirection(x, y, +1, +1, seatType))
	return count
}

func (m seatMap) CountAdjacentSeatsOfType(x, y int, seatType seatType) int {
	count := 0
	count += btoi(m.Get(x-1, y-1) == seatType)
	count += btoi(m.Get(x+0, y-1) == seatType)
	count += btoi(m.Get(x+1, y-1) == seatType)
	count += btoi(m.Get(x-1, y+0) == seatType)
	count += btoi(m.Get(x+1, y+0) == seatType)
	count += btoi(m.Get(x-1, y+1) == seatType)
	count += btoi(m.Get(x+0, y+1) == seatType)
	count += btoi(m.Get(x+1, y+1) == seatType)
	return count
}

func (m seatMap) HasSeatVisibleInDirection(x, y, dx, dy int, seatType seatType) bool {
	x += dx
	y += dy
	switch m.Get(x, y) {
	case seatType:
		return true
	case seatTypeFloor:
		return m.HasSeatVisibleInDirection(x, y, dx, dy, seatType)
	}
	return false
}

func (m seatMap) Get(x, y int) seatType {
	if x < 0 || y < 0 || x >= m.RowWidth {
		return seatTypeNone
	}
	i := y*m.RowWidth + x
	if i >= len(m.Seats) {
		return seatTypeNone
	}
	return m.Seats[i]
}

func (m seatMap) RearrangeUntilStable(scratch [2]seatMap, rearrange func(_ seatMap, x, y int, _ seatType) seatType) seatMap {
	scratchCounter := 0
	src := &m
	dst := &scratch[scratchCounter]
	for {
		changes := 0
		for i, st := range src.Seats {
			x := i % src.RowWidth
			y := i / src.RowWidth
			newST := rearrange(*src, x, y, st)
			dst.Seats[i] = newST
			if newST != st {
				changes++
			}
		}
		if changes == 0 {
			return *dst
		}
		scratchCounter ^= 1
		dst, src = &scratch[scratchCounter], dst
	}
}

func (m seatMap) CountSeatsOfType(seatType seatType) int {
	count := 0
	for _, st := range m.Seats {
		if st == seatType {
			count++
		}
	}
	return count
}

func (m seatMap) Clone() seatMap {
	seats := make([]seatType, len(m.Seats))
	copy(seats, m.Seats)
	return seatMap{m.RowWidth, seats}
}

type seatType byte

const (
	seatTypeNone seatType = iota
	seatTypeFloor
	seatTypeEmpty
	seatTypeTaken
)

func btoi(b bool) int {
	if b {
		return 1
	}
	return 0
}
